### PR TITLE
Top P, K, P+K CUDA Kernel

### PR DESCRIPTION
**Pull request by aciddelgado**
Accelerate Top P and Top K sampling on GPU with CUDA. Integrate these so it is easier to change them.

Average performance numbers using: vocab_size=32000, batch_size=1 on RTX 4090
Average time taken by TopP: 58.87 microseconds (p=0.95)
Average time taken by TopK: 28.07 microseconds (k=5)
Average time taken by TopP+K: 28.56 microseconds (k=5, p=0.95)

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,74 @@
+{
```
Comment (RyanUnderhill) [inline .vscode/settings.json:1]: I don't think you meant to make this part of your PR

File: build.sh [hunk 2]
```diff
@@ -9,6 +9,6 @@ export CUDA_HOME="/usr/local/${cuda_version}"
 export CUDNN_HOME="/usr/lib/x86_64-linux-gnu/"
 cuda_compiler="/usr/local/${cuda_version}/bin/nvcc"
 
-cmake -S . -B build -DCMAKE_CUDA_ARCHITECTURES=$cuda_arch -DCMAKE_CUDA_COMPILER=$cuda_compiler -DCMAKE_POSITION_INDEPENDENT_CODE=ON -DUSE_CXX17=1 -DUSE_CUDA=1
+cmake -S . -B build -DCMAKE_CUDA_ARCHITECTURES=$cuda_arch -DCMAKE_CUDA_COMPILER=$cuda_compiler -DCMAKE_POSITION_INDEPENDENT_CODE=ON -DUSE_CXX17=1 -DUSE_CUDA=1 -G Ninja
 cd build
-make
```
Comment (RyanUnderhill) [inline build.sh:14]: Any drawbacks to Ninja vs: `cmake --build . -j $(nproc)` ?

File: install_ort.sh [hunk 3]
```diff
@@ -2,5 +2,5 @@
```
Comment (RyanUnderhill) [inline install_ort.sh:2]: Shouldn't be part of your PR?

File: src/search_cuda.cpp [hunk 4]
```diff
@@ -83,6 +86,17 @@ void Search_Cuda::SetLogits(RoamingArray<float> logits_unk) {
 
     cuda::Launch_log_softmax(target.data(), static_cast<int>(target.size()), params_.cuda_stream);
   }
+
```
Comment (RyanUnderhill) [inline src/search_cuda.cpp:89]: Commented out code?

File: src/tests/tests.cpp [hunk 5]
```diff
@@ -235,6 +236,122 @@ void Test_BeamSearch_Gpt_Cuda() {
     Test_BeamSearch_Gpt_Cuda(model_path.first, model_path.second);
 }
 
+#include "tests_helper.cuh"
+void Test_TopP_Cuda() {
+  std::cout << "Test_TopP" << std::flush;
+
+  //////////// Simple Test
```
Comment (RyanUnderhill) [inline src/tests/tests.cpp:243]: Too many slashes

**Comment by yufenglee** [comment]
could you please add some perf numbers?

File: src/search.h [hunk 6]
```diff
@@ -23,6 +23,7 @@ struct Search {
   virtual void SelectTop() = 0;
   virtual void SampleTopP(float /*p*/, float /*temperature*/) { assert(false); }
   virtual void SampleTopK(int /*k*/, float /*temperature*/) { assert(false); }
+  virtual void GetTopKSubset(int* /*tokens_out*/, int /*k*/) { assert(false); }
```
Comment (yufenglee) [inline src/search.h:26]: >GetTopKSubset [](http://example.com/codeflow?start=15&length=13)

what does subset mean?

Comment (yufenglee) [inline src/search.h:26]: >assert(false) [](http://example.com/codeflow?start=63&length=13)

add an error message like: TopK is only implemented in SamplingSearch.

File: src/search_cuda.h [hunk 7]
```diff
@@ -48,8 +48,10 @@ struct GreedySearch_Cuda : Search_Cuda {
   RoamingArray<int32_t> GetNextIndices() override { return gpu_span<int32_t>{}; }
 
   void SelectTop() override;
-  void SampleTopK(int k, float t) override { assert(false); }
+  void SampleTopK(int k, float t) override;
```
Comment (yufenglee) [inline src/search_cuda.h:51]: why implement SampleTopK in GreedySearch?

Comment (yufenglee) [reply src/search_cuda.h:51]: I think we may refactor the search into two different class: Search and BeamSearch. Search supports function: SelectTop, SelectTopK, SelectTopP, SelectTopKTopP. @RyanUnderhill

File: src/search_cuda.cpp [hunk 8]
```diff
@@ -164,48 +166,78 @@ void GreedySearch_Cuda::SelectTop() {
...
-  // Find the first token where the cumulative probability exceeds the threshold
-  for (int i = 0; i < scores.size(); i++) {
-    threshold -= scores[indices[i]];
-    if (threshold > 0)
-      continue;
-
-    token = indices[i];
-    break;
-  }
+// TODO: commented out in case of benchmarking
```
Comment (yufenglee) [inline src/search_cuda.cpp:169]: > TODO: commented out in case of benchmarking [](http://example.com/codeflow?start=2&length=44)

do we need this function? if we need, we should uncomment it, or we need to remove those code

Comment (yufenglee) [reply src/search_cuda.cpp:169]: if you just want to use it as a reference of performance, we don't need it after benchmark.

File: src/cuda_sampling.cu [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+#include <cub/cub.cuh>
+#include <curand_kernel.h>
+#include <iostream>
+
+namespace Generators {
+namespace cuda {
+
+const int max_threads = 1024;
+constexpr int GPU_WARP_SIZE = 32;
+constexpr int GPU_WARP_SIZE_HOST = 32;
```
Comment (yufenglee) [inline src/cuda_sampling.cu:22]: what is the difference between those 2 variables?

File: src/cuda_sampling.cu [hunk 10]
```diff
@@ -0,0 +1,584 @@
...
+};
+
+// aligned vector generates vectorized load/store on CUDA
+template <typename T, int vec_size>
+struct alignas(sizeof(T) * vec_size) aligned_vector {
+  T val[vec_size];
+};
+
+template <template <typename, typename> class Reduction, int ILP, typename T, typename AccumT>
+__device__ __forceinline__ AccumT ilpReduce(int shift, T* data, int size, const Reduction<T,
```
Comment (yufenglee) [inline src/cuda_sampling.cu:68]: >ilpReduce [](http://example.com/codeflow?start=34&length=9)

please follow the convention for function naming

File: src/cuda_sampling.cu [hunk 11]
```diff
@@ -0,0 +1,584 @@
+// Filter Logits Kernel implmentation from ONNX Runtime
```
Comment (yufenglee) [inline src/cuda_sampling.cu:1]: >/ Filter Logits Kernel implmentation from ONNX Runtime [](http://example.com/codeflow?start=1&length=54)

add copyright

File: src/cuda_sampling.cu [hunk 12]
```diff
@@ -0,0 +1,584 @@
...
+namespace Generators {
+namespace cuda {
+
+const int max_threads = 1024;
+constexpr int GPU_WARP_SIZE = 32;
+constexpr int GPU_WARP_SIZE_HOST = 32;
+
+////////////// SOFTMAX KERNELS AND LAUNCHER //////////////
+
+// TODO: why do we need all these template struct things, can't we just do this in the main kernel???
```
Comment (yufenglee) [inline src/cuda_sampling.cu:26]: >TODO: why do we need all these template struct things, can't we just do this in the main kernel??? [](http://example.com/codeflow?start=3&length=98)

it is for code sharing.

File: src/cuda_sampling.cu [hunk 13]
```diff
@@ -0,0 +1,605 @@
...
+#include "smartptrs.h"
+#include <cuda_runtime.h>
+#include <cub/cub.cuh>
+#include <curand_kernel.h>
+#include <iostream>
+
+namespace Generators {
+namespace cuda {
+
+const int max_threads = 1024;
```
Comment (yufenglee) [inline src/cuda_sampling.cu:20]: >const int [](http://example.com/codeflow?start=0&length=9)

nit: constexpr

Comment (yufenglee) [inline src/cuda_sampling.cu:20]: >max_threads [](http://example.com/codeflow?start=10&length=11)

nit: name it as kMaxThreads

File: src/cuda_sampling.cu [hunk 14]
```diff
@@ -0,0 +1,605 @@
...
+constexpr int GPU_WARP_SIZE = 32;
```
Comment (yufenglee) [inline src/cuda_sampling.cu:21]: nit: similar here. kGPUWarpSize

File: src/cuda_sampling.cu [hunk 15]
```diff
@@ -0,0 +1,605 @@
...
+    LaunchGetTopKSubsetFullSort(stream, scores_softmaxed.data(), scores_out, indices_out, vocab_size, batch_size, k);
+  }
+}
+
+
+// Kernel launcher for combined (or seperate) top k and top p sampling; where k is the max number of tokens to sample and p is the probability threshold
+void GetSample(cudaStream_t stream, int32_t* next_token_out, float* scores_in, int vocab_size, int batch_size, int k, float p, float temperature) {
+  int sample_range = (k > 0 && k < vocab_size) ? k : vocab_size;
+  auto scores_buffer = CudaMallocArray<float>(sample_range * batch_size);
+  auto indices_buffer = CudaMallocArray<int>(sample_range * batch_size);
```
Comment (yufenglee) [inline src/cuda_sampling.cu:592]: malloc and free synchronize the device. We need to create this buffer at the beginning of the decoding and share it across for the whole decoding stage.

File: src/cuda_sampling.cu [hunk 16]
```diff
@@ -0,0 +1,605 @@
...
+  } else if (k <= 8) {
+    GetTopK(8);
+  } else if (k <= 16) {
+    GetTopK(16);
+  } else if (k <= 32) {
+    GetTopK(32);
+  } else if (k <= 64) {
+    GetTopK(64);
+  } else {
+    LaunchGetTopKSubsetFullSort(stream, scores_softmaxed.data(), scores_out, indices_out, vocab_size, batch_size, k);
```
Comment (yufenglee) [inline src/cuda_sampling.cu:583]: >scores_softmaxed [](http://example.com/codeflow?start=40&length=16)

scores_softmaxed is already SoftMax. SoftmaxAndSort invoked in LaunchGetTopKSubsetFullSort will do softmax again.

File: src/cuda_sampling.cu [hunk 17]
```diff
@@ -0,0 +1,605 @@
...
+  }
+  SampleKernel<256><<<grid, block, 0, stream>>>(prefix_sums.data(), indices, index_out, sample_range, thresholds.data());
+}
+
+/////////////// TOP P+K KERNEL LAUNCHER ////////////////////
+
+// Outputs sorted scores and corresponding indices... scores_out and indices_out should already be allocated
+void SoftmaxAndSort(cudaStream_t stream, float* scores_in, float* scores_out, int* indices_out, int vocab_size, int batch_size, float temperature) {
+  // Softmax scores
+  auto scores_buffer = CudaMallocArray<float>(vocab_size * batch_size);
```
Comment (yufenglee) [inline src/cuda_sampling.cu:524]: >CudaMallocArray [](http://example.com/codeflow?start=23&length=15)

and all those memory need to be allocated ahead of time

**Comment by yufenglee** [comment]
I think you can merge your PR firstly and have a following PR to fix the double softmax and memory sharing issue.

File: src/cuda_sampling.cu [hunk 18]
```diff
@@ -0,0 +1,605 @@
...
+
+  float k_prob = prefix_sums[k-1];
+  if (index < batch_size) {
+    float min_p = fminf(p, k_prob);
+    thresholds[index] = min_p * curand_uniform(&state);
+  }
+}
+
+// Sets up random thresholds for top p or top k sampling
+__global__ void RandomThresholdKernelTopP(int seed, float* thresholds, float* prefix_sums, int batch_size, float p) {
```
Comment (yufenglee) [inline src/cuda_sampling.cu:445]: >prefix_sums [](http://example.com/codeflow?start=78&length=11)

does the unused parameter cause build break?

**Review by yufenglee — APPROVED** [review]

File: test/CMakeLists.txt [hunk 19]
```diff
@@ -31,8 +31,13 @@ target_link_libraries(unit_tests PRIVATE
   GTest::gtest_main
 )
 if(USE_CUDA AND CMAKE_CUDA_COMPILER)
+  file(GLOB cuda_test_srcs CONFIGURE_DEPENDS
+    "${CMAKE_CURRENT_SOURCE_DIR}/*.cu"
+    "${CMAKE_CURRENT_SOURCE_DIR}/*.cuh"
+  )
   set_target_properties(unit_tests PROPERTIES LINKER_LANGUAGE CUDA)
-  target_link_libraries(unit_tests PRIVATE cublasLt cublas cudnn curand cufft cudart)
+  # target_link_libraries(unit_tests PRIVATE cublasLt cublas cudnn curand cufft cudart)
```
Comment (yufenglee) [inline test/CMakeLists.txt:39]: nit: remove this line
